### ISSUE TITLE
user/password proxy support for google cloud storage client

### ISSUE THREAD

**Issue by Eliyahu-Machluf** [googleapis/java-storage#125]
Hi,

I'm writing a client for google cloud storage using the java jar google-cloud-storage version 1.37.1
maven excerpt 
 <dependency>
            <groupId>com.google.cloud</groupId>
            <artifactId>google-cloud-storage</artifactId>
            <version>1.37.1</version>
        </dependency>

I want to configure it work with proxy which uses username and password, but the code does not work. 
I configure my code to use proxy using System.setProperty method
                System.setProperty("http.proxyUser", my_username);
                System.setProperty("http.proxyPassword", my_password);

However, this does not work.
If I work with proxy without username/password, it works fine.
Also using proxy which expect to get SSL connections, does not work.

**My question is:**
1. What shall I do to enable working with proxy with username/password, and with SSL.
2. I know I'm using an old jar, if I upgrade it to the latest one, will this give me support for proxy with user/password?

Thanks.

Eliyahu

**Comment by dmitry-fa** [comment]
Hi @Eliyahu-Machluf,

Thank you for your question. 
Could you check if setting https.proxyUser / https.proxyPassword hepls

**Comment by dmitry-fa** [comment]
quoting: [Java Networking and Proxies](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html)

The https (http over SSL) protocol handler has its own set of properties:
    https.proxyHost
    https.proxyPort

**Comment by Eliyahu-Machluf** [comment]
I configure both http.proxyUser,http.proxyPassword and https.proxyUser,https.proxyPassword,
**But it does not work.**
Regarding the cite from 'Java Networking and Proxies', I want my http request be tunneled through proxy, and connect with the proxy using SSL.

**To clarify:** 
My proxy support the following ways to connect with it.
1. http 
2. http+ user/password
3. https
4. https+user/password.

I want my google cloud storage client to be able to connect the proxy at all 4 ways. currently only the first one works.

**Comment by dmitry-fa** [comment]
@Eliyahu-Machluf, could you confirm that your proxy server is configured to meet requirements from [Google Storage > Documentation > Support > Proxy Servers](https://cloud.google.com/storage/docs/troubleshooting#proxy-server)

If it's configured properly, could you provide more details about the errors you see working with your proxy.

**Comment by Eliyahu-Machluf** [comment]
Hi,

I validated the proxy is configured well, by using curl and executing some get command (loading pages from the web) using proxy.

We have proxy with 4 configurations, listening on different ports:
1. http
2. http and user/password
3. https
4. https and user/password.

All configurations were checked using curl, and are all fine. 
However, when I create my google cloud storage client, only the first configuration (http without user/password) works. Other configurations fails.

Here is a code snippet with the relevant code:
```
        System.setProperty("http.proxyHost", proxyHost);
        System.setProperty("http.proxyPort", String.valueOf(proxyPort));
        System.setProperty("https.proxyHost", proxyHost);
        System.setProperty("https.proxyPort", String.valueOf(proxyPort));
        if (proxyUser.length() > 0) {
     //System.out.println("Configured proxy user” + proxyUser );
            System.setProperty("http.proxyUser", proxyUser);
            System.setProperty("https.proxyUser", proxyUser);
        }
        if (proxyPassword.length() > 0) {
     //System.out.println("Configured proxy password” + proxyPassword );
            System.setProperty("http.proxyPassword", proxyPassword);
            System.setProperty("https.proxyPassword", proxyPassword);
        }

        GoogleCredentials credentials;
        try {
            credentials = GoogleCredentials.fromStream(new ByteArrayInputStream(googlestorageConfiguration.getJsonCredentials().getBytes()))
                    .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform"));
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage());
        }
        storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
        bucket = storage.get(bucketName);
```

**The last line (bucket=storage.get(bucketName) fails when code is configured to use proxy of user/password with http protocol:**
When trying to connect proxy http protocol with user/password, I get the following warning:

 17, 2020 2:19:10 PM com.google.api.client.http.HttpRequest execute
WARNING: exception thrown while executing request
java.io.IOException: Unable to tunnel through proxy. **Proxy returns "HTTP/1.1 407 Proxy Authentication Required"**
        at sun.net.www.protocol.http.HttpURLConnection.doTunneling(Unknown Source)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(Unknown Source)
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
        at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:383)
        at com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:175)
        at com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:161)
        at com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:96)
        at com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:161)
        at com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:120)
        at com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:93)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:300)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
        at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:379)
        at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:176)
        at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:173)
        at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:89)
        at com.google.cloud.RetryHelper.run(RetryHelper.java:74)
        at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:51)

and finally the error “Error getting access token for service account”
 [FILE_FACTORY    ]T:  JAVA EXCEPTION: com.google.cloud.storage.StorageException: Error getting access token for service account: 
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:220)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:382)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:176)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:173)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:89)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:74)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:51)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:172)
	at com.attunity.AttGCPStorageFileFactory.doSetParams(AttGCPStorageFileFactory.java:364)
	at com.attunity.FileFactoryLoggerDecorator.setParams(FileFactoryLoggerDecorator.java:78)
	at com.attunity.AttGCPStorageFileFactory.setParams(AttGCPStorageFileFactory.java:74)
Caused by: java.io.IOException: Error getting access token for service account: 
	at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:385)
	at com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:175)
	at com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:161)
	at com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:96)
	at com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:161)
	at com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:120)
	at com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:93)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:300)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:379)
	... 9 more
Caused by: java.io.IOException: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Authentication Required"
	at sun.net.www.protocol.http.HttpURLConnection.doTunneling(Unknown Source)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(Unknown Source)
	at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
	at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:383)
	... 20 more  (AttLogger.java)

When configured to work with proxy **https** port I get the following warning
??? 17, 2020 2:04:37 PM com.google.api.client.http.HttpRequest execute
WARNING: exception thrown while executing request
java.net.SocketException: **Connection reset**
        at java.net.SocketInputStream.read(Unknown Source)
        at java.net.SocketInputStream.read(Unknown Source)
        at java.io.BufferedInputStream.fill(Unknown Source)
        at java.io.BufferedInputStream.read1(Unknown Source)
        at java.io.BufferedInputStream.read(Unknown Source)
        at sun.net.www.http.HttpClient.parseHTTPHeader(Unknown Source)
        at sun.net.www.http.HttpClient.parseHTTP(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.doTunneling(Unknown Source)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(Unknown Source)
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
        at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:383)
        at com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:175)
        at com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:161)
        at com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:96)
        at com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:161)
        at com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:120)
        at com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:93)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:300)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
        at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:379)
        at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:176)
        at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:173)
        at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:89)
        at com.google.cloud.RetryHelper.run(RetryHelper.java:74)
        at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:51)
        at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:172)

And finally an exception failed to get access token:

 com.google.cloud.storage.StorageException: **Error getting access token for service account**: 
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:220)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:382)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:176)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:173)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:89)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:74)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:51)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:172)
	at com.attunity.AttGCPStorageFileFactory.doSetParams(AttGCPStorageFileFactory.java:364)
	at com.attunity.FileFactoryLoggerDecorator.setParams(FileFactoryLoggerDecorator.java:78)
	at com.attunity.AttGCPStorageFileFactory.setParams(AttGCPStorageFileFactory.java:74)
Caused by: java.io.IOException: Error getting access token for service account: 
	at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:385)
	at com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:175)
	at com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:161)
	at com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:96)
	at com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:161)
	at com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:120)
	at com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:93)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:300)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:379)
	... 9 more
Caused by: java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at java.io.BufferedInputStream.fill(Unknown Source)
	at java.io.BufferedInputStream.read1(Unknown Source)
	at java.io.BufferedInputStream.read(Unknown Source)
	at sun.net.www.http.HttpClient.parseHTTPHeader(Unknown Source)
	at sun.net.www.http.HttpClient.parseHTTP(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.doTunneling(Unknown Source)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(Unknown Source)
	at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
	at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:383)

**Comment by Eliyahu-Machluf** [comment]
But before digging into the exceptions, I have a simple question:
Does the java client for google cloud storage support proxy with user/password, and proxy with https connection?

**Comment by dmitry-fa** [comment]
> But before digging into the exceptions, I have a simple question:
> Does the java client for google cloud storage support proxy with user/password, and proxy with https connection?

Hi, I think google storage works fine with proxies, at least I've never heard opposite. I believe your problems could be fixed by proper configuration. I'll try to help you out.

**Comment by dmitry-fa** [comment]
> java.io.IOException: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Authentication Required"
As recommended in [Oracle 8u111 Update Release Notes](https://www.oracle.com/technetwork/java/javase/8u111-relnotes-3124969.html)

try run java with: 
`-Djdk.http.auth.tunneling.disabledSchemes=`
or add to your code:
`System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");`

**Comment by Eliyahu-Machluf** [comment]
I tried adding `System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");` statement to the code, but this did not solve the issue, i get the same errors when using proxy with http+user/password, or when using proxy with https.

**Comment by dmitry-fa** [comment]
Hi @Eliyahu-Machluf ,

Sad to hear that disabling tunneling tunneling didn't help

Regarding:
> StorageException: Error getting access token for service account

It looks like the storage server doesn't get token from your request.
Quoting [Cloud Storage authentication > API authentication](https://cloud.google.com/storage/docs/authentication#apiauth):

> If your requests are being routed through a proxy, you may need to check with your network administrator to ensure that the Authorization header containing your credentials is not stripped out by the proxy.

Because neither storage client nor server do anything proxy specific, I think, the root cause should be searched within the proxy parameters. To eliminate java from the path you can check if `gsutil` (the command line utility to deal with storage)  works fine behind the proxy.
The following links can help you to install and setup gsutil to work with proxy:
[Install gsutil](https://cloud.google.com/storage/docs/gsutil_install)
[Configuring gcloud, gsutil and bq to use Proxy Servers](https://medium.com/google-cloud/configuring-gcloud-gsutil-and-bq-to-use-proxy-servers-4f09dbaf42c2)

**Comment by Eliyahu-Machluf** [comment]
Hi Dmitry,

Thanks for your assistance. Currently I'm busy with other tasks, I hope to have time at next week to check the proxy user/password with `gsutil` utility.

**Comment by dmitry-fa** [comment]
Hi Eliyahu, 
I'll be glad to help. Please keep me informed of the progress.
Thanks

**Comment by Eliyahu-Machluf** [comment]
Hi Dmitry,

I had some time today and I did some additional tests:
1. I've updated the Google Cloud Storage jar I use to version 1.105.2 (from March 2020), and tested it again. It gave the same results: with http proxy without user/password it worked. with user/password, and with https proxy it did not work.

2. As per your suggestion, I installed the gcloud utility CLI, configured it and run gsutil session. I succeeded to upload file using proxy http with and without user/password.
I don't know how to configure it to run with https proxy. as I cannot configure gcloud to run with https proxy. running the command:
gcloud config set proxy/type **https**
gives error:
ERROR: (gcloud.config.set) The proxy type property value [https] is not valid. Possible values: [http, http_no_tunnel, socks4, socks5].

To summarize:
- http proxy with user/password works with gsutil CLI, and does not work with the Java SDK.
- https proxy - I don't know whether this is supported by gsutil CLI, and the Java SDK fails to connect using this configuration.

I want to ask again: Does the Java SDK support http proxy with user/password and https proxy? As it seems it doesn't

Thanks.

Eliyahu

**Comment by dmitry-fa** [comment]
Hi Eliyahu, 

Exactly the same case is described here: https://github.com/forcedotcom/wsc/issues/87
Would you try to start your java with: `-Djdk.http.auth.tunneling.disabledSchemes=""`

Thanks,
Dmitry

**Comment by Eliyahu-Machluf** [comment]
Hi Dmitry,

I've added my code the line `System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");`
But still I get an exception: **com.google.cloud.storage.StorageException: Error getting access token for service account: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Authentication Required"**
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:229)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:406)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:217)
	at com.google.cloud.storage.StorageImpl$4.call(StorageImpl.java:214)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:213)
....
Caused by: java.io.IOException: Error getting access token for service account: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Authentication Required"
	at com.google.auth.oauth2.ServiceAccountCredentials.refreshAccessToken(ServiceAccountCredentials.java:444)
	at com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:157)
	at com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:145)
	at com.google.auth.oauth2.ServiceAccountCredentials.getRequestMetadata(ServiceAccountCredentials.java:603)
	at com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:91)
	at com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:159)
	at com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:109)
	at com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:88)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:423)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:542)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:475)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:592)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:403)
	... 9 more

**Comment by dmitry-fa** [comment]
What Java version do you use?

**Comment by Eliyahu-Machluf** [comment]
Hi Dmitry,

I use JRE 1.8.0_211, running on windows Server 2012

**Comment by dmitry-fa** [comment]
Would you try to start your java with: `-Djdk.http.auth.tunneling.disabledSchemes=""`
Your System.setProperty() might be invoked too late and have no effect. JDK library classes are probably already initialized by that moment.

**Comment by Eliyahu-Machluf** [comment]
I've tried adding starting the JVM with the option `-Djdk.http.auth.tunneling.disabledSchemes=""`, but it did not help. fails with the same exception

**Comment by dmitry-fa** [comment]
Another possible reason why this option doesn't help, the VM running code to access external resource through proxy is launched separately: you start one application passing extra JVM option and this JVM starts another one without that option.

**Comment by dmitry-fa** [comment]
Hi Eliyahu,

I installed proxy locally and managed to reproduce your failure.
It's much easier for me to find a solution now.

**Comment by Eliyahu-Machluf** [comment]
Great.
I'm happy to hear. 

Note that Java has a default behavior that if it fails to connect a host using the proxy, it will try to connect directly to that host. So I reproduced the issue, with a machine which does not have direct access to the internet, but only using a proxy.

**Comment by dmitry-fa** [comment]
Good news, I managed to get on through proxy, the following lines are required:
```
        System.setProperty("http.proxyHost", PROXY_HOST);
        System.setProperty("http.proxyPort", PROXY_PORT);
        System.setProperty("https.proxyHost", PROXY_HOST);
        System.setProperty("https.proxyPort", PROXY_PORT);
        System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
        Authenticator.setDefault(new Authenticator() {
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(PROXY_USER, PROXY_PASS.toCharArray());
            }
        });
```

Below is the entire program:
```
package proxy;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.Authenticator;
import java.net.PasswordAuthentication;
import java.net.URL;

public class Test {
    private static final String PROXY_HOST = "10.10.10.10";
    private static final String PROXY_PORT = "3128";
    private static final String PROXY_USER = "user";
    private static final String PROXY_PASS = "1234";

    public static void main(String... args) throws Exception {
        System.setProperty("http.proxyHost", PROXY_HOST);
        System.setProperty("http.proxyPort", PROXY_PORT);
        System.setProperty("https.proxyHost", PROXY_HOST);
        System.setProperty("https.proxyPort", PROXY_PORT);
        System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
        Authenticator.setDefault(new Authenticator() {
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(PROXY_USER, PROXY_PASS.toCharArray());
            }
        });
        URL url = new URL("https://example.com/");
        try (BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()))) {
            String inputLine;
            while ((inputLine = in.readLine()) != null)
                System.out.println(inputLine);

        }
    }
}
```

**Comment by Eliyahu-Machluf** [comment]
Hi Dmitry,

These are great news. Thanks for the effort.
I copied the changes to my test utility and managed to work with http proxy with user/password.
Great!

I still cannot work with http**s** proxy. I use Squid proxy, and it is configured also to be accessed using https connection. These connections does not work with and without user/password.

If you have a working environment can you test this as well?

Thanks.

**Comment by dmitry-fa** [comment]
The code I provided works with both: http and https. You can access http://example.com as well as https://example.com
 Storage is also accessible through the proxy.

**Comment by dmitry-fa** [comment]
Hi Eliyahu,
I'm afraid I can't help you with establishing HTTPS connection between your client and proxy.
Probably, the following link would be helpful for you:
https://stackoverflow.com/questions/10440690/pros-and-cons-of-using-a-http-proxy-v-s-https-proxy/10442767

I guess, the thing you are looking for is called: Connecting to a proxy over SSL/TLS.
Are you sure that you need this?

**Comment by dmitry-fa** [comment]
Hi @Eliyahu-Machluf ,

Have you managed to resolve your problems with proxy?
Can we close this issue?

**Comment by Eliyahu-Machluf** [comment]
Hi @dmitry-fa,

Thanks for reproducing the problem, and showing how to solve it.

I'm closing the issue.

**Comment by dmitry-fa** [comment]
Hi @Eliyahu-Machluf,

The new documentation has been just published:
https://github.com/googleapis/google-auth-library-java/blob/master/README.md#configuring-a-proxy

I hope it will help you to setup an authenticated proxy.

**Comment by Eliyahu-Machluf** [comment]
Great.
Thanks.
